### PR TITLE
Feature: external sources

### PR DESCRIPTION
**Pull request by KevinVdV**
This is the REST API development for DSpace/Rest7Contract#74

In this PR we introduce the new concept "external sources". These external sources will be the backbone for authority control & metadata imports using third party API's.

You can interact with external sources by using the "ExternalDataService" interface. The implementation of this interface is just a wrapper around a list of ExternalDataProvider objects. The ExternalDataProvider is an interface that offers search methods (find one / find by query), the implementations of these interfaces are the ones that are actually doing to the calls to the third party sources (SHERPA/RoMEO, ORCID, Library of Congress, ...) 
The ExternalDataProvider implementations are configured in the config/spring/api/external-services.xml file.

The ExternalDataProvider will return results in the generic "ExternalDataObject" class.

### CODE REVIEW

**Comment by paulo-graca** [comment]
This PR it's working as expected witch regards the RestContract. But I think it will be useful to have the Entry's context at EntryValue level, so I've created this PR: https://github.com/DSpace/Rest7Contract/pull/88 to address that need.

File: dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java]: Should we anticipate service unavailability and throw a generic external source exception?

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java]: I think this way of mapping Orcid, in the code, is less flexible and might be limiting access to other relevant information like the person's identifiers (scopus, researcher id, etc...), or author's name variants. I think XSL transformations could be used here, with the output of DSpace Intermediate Metadata: https://wiki.duraspace.org/display/DSPACE/DSpaceIntermediateMetadata

Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java]: If relevant, it can be addressed in another Pull Request

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java]: We have cases when the submitter knows the ORCID ID of the author. In those cases it's useful to ignore que ORCID search and simply do the record fetch. Otherwise, the query will return millions of results and we will do unnecessary requests, resulting in a delayed output for the client.

If we identify that the query value is an ORCID ID, then we firstly do simple checkdigit validation to ensure we have a valid ORCID ID. And if valid, then perform the record fetch.

How it works:
**ORCID-ID:** 0000-0003-1234-4321
**stripped(ORCID-ID):** 0000000312344321
**baseDigits:** 000000031234432 
**generateOrcidCheckDigit(baseDigits):** 1
**(stripped(ORCID-ID).equals(baseDigits+generateOrcidCheckDigit(baseDigits)))**: true


```

      /* Generates check digit as per ISO 7064 11,2.
       *
       */
       private static String generateOrcidCheckDigit(String baseDigits) {
             int total = 0;
             for (int i = 0; i < baseDigits.length(); i++) {
                    int digit = Character.getNumericValue(baseDigits.charAt(i));
                    total = (total + digit) * 2;
             }
             int remainder = total % 11;
             int result = (12 - remainder) % 11;
             return result == 10 ? "X" : String.valueOf(result);
       }
```

It can be considered in another PR as an improvement of Orcid Person provider.

**Review by paulo-graca — COMMENTED** [review]
Overall, this PR works as expected. I've some improvements proposal that can be address in different and subsequent pull requests.
One of those improvements is related with the unavailability of the external source service. I've done some tests and forced an error and got an Internal Server Error on the REST payload. Should the client interpret this error or should a generic error be thrown?
Thanks for this @KevinVdV and others!

File: dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java]: A generic 500 error deals with any potential issue (doesn’t specify any details)
A 404 assumes the specific resource doesn’t exist, but is not applicable if e.g. the connection to ORCID failed
To avoid over-complicating potential reasons why an external source is causing an error, I'd prefer to keep a generic error at least for the first implementation.
If there's a decent set of integrations, it can be relevant to verify if there are some common exceptions which can be treated separately

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV2AuthorDataProvider.java]: That's a good suggestion, but not really the scope of this PR. The goal here is to have the framework in place and a sample implementation.
This can best be handled in a separate ticket

File: dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/external/provider/ExternalDataProvider.java]: A generic 500 error deals with any potential issue (doesn’t specify any details)
A 404 assumes the specific resource doesn’t exist, but is not applicable if e.g. the connection to ORCID failed
To avoid over-complicating potential reasons why an external source is causing an error, I'd prefer to keep a generic error at least for the first implementation.
If there's a decent set of integrations, it can be relevant to verify if there are some common exceptions which can be treated separately

**Review by paulo-graca — APPROVED** [review]
Thank you @KevinVdV and @benbosman !

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java]: I see a lot of (conceptual) similarities between the `/api/integrations/externalsources` endpoint (represented here) and the existing `/api/integrations/authorities` endpoint.  So, I'm slightly surprised that this `ExternalSourceRestRepository` (and the corresponding Controller) seems to take a different implementation approach to that of `AuthorityRestRepository`, `AuthorityEntryLinkRepository` and `AuthorityEntryValueLinkRepository` (the three of which work together so there is no `AuthorityRestController`).   Could you review how these Authority classes implement something similar and see if it can be applied here?   I worry we are building a lot of reliance on custom Controllers to override default behaviors, when we could be using RestRepository (and LinkRestRepository) classes to do the same thing.

I'd like us to follow existing practices where possible. Since these endpoints behave similarly, I wonder if we can use the same approach to both?

File: dspace/config/dspace.cfg
Comment (tdonohue) [inline dspace/config/dspace.cfg]: I don't think we need to remove these configurations from our dspace.cfg.  Please remember, that in Spring Bean definitions we can *reference* these configurations using `${config}` references.  Please take a look at how we do this in our `core-hibernate.xml` to pull in the database settings from dspace.cfg: https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/core-hibernate.xml#L41-L49

I'd rather make these configurations EASY to configure and override (in your local.cfg).  So I think we need to keep these configurations in dspace.cfg instead of moving them into Spring XML configs (which are more difficult to edit & override).

File: dspace/config/dspace.cfg
Comment (tdonohue) [inline dspace/config/dspace.cfg]: Same here. Don't remove ORCID configurations from dspace.cfg please.  They should stay here and be referenced from the Spring Beans that need them.

File: dspace/config/spring/api/external-services.xml
Comment (tdonohue) [inline dspace/config/spring/api/external-services.xml]: As noted above all `<property>` settings in this XML file should be pulling in values from dspace.cfg.  Otherwise we are making these settings more difficult to edit/change (as DSpace installers couldn't easily override them in a local.cfg, unless they are in dspace.cfg)

**Review by tdonohue — COMMENTED** [review]
@KevinVdV : I gave this a code review today. Mostly it's looking good, but I do have some (mostly minor) implementation concerns in inline comments below.

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAResponse.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAResponse.java]: Setting aside the possibility of a null message, I suggest that code should always try to explain what has happened *in the context of that code*.  When the callee throws an exception, it doesn't know what that means to the caller, but the caller does and can better explain the consequences of the abrupt completion.

File: dspace-api/src/main/java/org/dspace/mock/MockMetadataValue.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/mock/MockMetadataValue.java]: Would `DTO` or `Holder` be a useful term here?  This looks like an object that is supposed to carry around the content of a persistence entity without being a persistence entity.

File: dspace/config/dspace.cfg
Comment (mwoodiupui) [reply dspace/config/dspace.cfg]: would like to participate in a more general discussion, elsewhere, of how we configure DSpace.

While @tdonohue 's concerns are reasonable, I wonder how likely it is that this will ever be touched, unless the remote service changes (in which case, we will *all* have to touch it).  There is something to be said for making things that normally shouldn't be adjusted, hard to mess up.

Comment (tdonohue) [reply dspace/config/dspace.cfg]: @mwoodiupui : I agree with your logic overall, but I think my original comment was taken slightly out of context.  You are correct, we may want to argue that the SHERPA API *url* is unlikely to need overriding. However, there's also a SHERPA "apikey" that will *always need overriding* but is unfortunately buried in Spring configs here: https://github.com/DSpace/DSpace/pull/2560/files#diff-c1d29d61b7b458582c79b2ac9383b356R13

So, my comment above was more to say "Let's please not move configs (in general) into Spring XML files, when we can just as easily reference them where they exist in `dspace.cfg`"  I wasn't meaning to imply this only applies to the URL setting.

That said, honestly, we might also want to keep the `sherpa.romeo.url` in the dspace.cfg for now, simply because it's used *twice* in Spring configs.  It'd be much easier to change in one place (when necessary).

Comment (mwoodiupui) [reply dspace/config/dspace.cfg]: Ah, yes, API keys are site configuration, not developer configuration, and should be readily accessible by site maintainers.  And the corresponding URL may as well stay with it for now.

File: dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java]: Can we please change this package name too?  The word `mock` is again misleading. It could be changed to simply `org.dspace.content.dto`, or even just `org.dspace.dto` is fine too.  Either is better in my opinion.

File: dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java]: This TODO should be removed.  We might want to improve the JavaDocs on this class though to make it clear that this `MetadataValueDTO` is no longer just for crosswalks.   It's simply an "in-memory data transfer object (DTO) for the MetadataValue class".

File: dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java]: Update the classname here please

File: dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/mock/MetadataValueDTO.java]: Again, update classname

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/MockMetadataConverter.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/MockMetadataConverter.java]: This Converter should be renamed so that it no longer includes the term `Mock`.  Perhaps this should be renamed `MetadataValueDTOListConverter`, since this converts a List of MetadataValueDTO objects.

File: dspace/config/spring/api/external-services.xml
Comment (tdonohue) [inline dspace/config/spring/api/external-services.xml]: This should be `${sherpa.romeo.url}`, not the ORCID API URL

**Review by tdonohue — COMMENTED** [review]
@KevinVdV and @Raf-atmire : This is looking better. Thanks for the updates.  I've got a few more inline comments though (all minor) as a few new mistakes were accidentally introduced.

File: dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAResponse.java
Comment (KevinVdV) [reply dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAResponse.java]: @tdonohue We have improved the logging here.

@mwoodiupui I agree with you, but I think this is best done in follow up PR's. In our PR we just refactored the third party sources without making alterations to how the actual calls are made. For example the code in current master doesn't log the exception at all: https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAResponse.java#L165

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java
Comment (KevinVdV) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java]: @tdonohue I agree with the premise that we should attempt to limit the number of controllers here. But the way it is currently implemented in the authority rest side is not the best approach for this functionality. It's hard to read and has a high risk of breaking the functionality with changes. It will be hard for future developers to find out how it works in the first place.

**How it works now**

The AuthorityEntryLinkRepository has a single “query” method while the AuthorityEntryValueLinkRepository has a single “getResource” method. When using intellij to find the usages of these methods, none are found. So it isn’t immediately clear how these are called in the first place. After some searching we found that the following controller method is responsible for the call: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L777
Reflection is used here to find the correct repository, the actual method call depends on the attributes found in the rest class: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/model/AuthorityRest.java#L17
Should you rename one of the methods in these repositories, it would still compile your code, but you would encounter exceptions at runtime. This is something we should avoid.

**Moving forward**

Further decreasing the controllers is perhaps something to verify next since the first iteration of projections is in place, and the next step there can be a detailed analysis on reflection and the RestResourceController to support permissions, .... Once the RestResourceController is refactored, it can be much easier to work with e.g. beans which will allow developers to more easily find out how repositories are contacted. But this is something that we would have to do in a separate PR dedicated to that single refactor.

**Comment by KevinVdV** [comment]
@tdonohue We have processed your feedback, merged with latest master to add projection support. So this is ready for another review.

File: dspace/config/dspace.cfg
Comment (tdonohue) [inline dspace/config/dspace.cfg]: @KevinVdV : Another very tiny change.  Could we please revert to using this existing `sherpa.romeo.apikey` and restore the comments about that configuration?   It looks like we've changed the name of this field to `sherpa.romeo.apiKey` (notice the capital "K"), which will mean that existing configurations will no longer work and will need updating.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java]: Thanks for the detailed response here, @KevinVdV .  The direction you've chosen sounds fine to me, and I'm good with this PR's approach as it has been recently refactored. Marking this conversation as "resolved"

File: dspace/config/dspace.cfg
Comment (tdonohue) [reply dspace/config/dspace.cfg]: After the above tiny change is made ⬆️ , this PR seems good to me. I don't notice any other issues or concerns.

**Review by tdonohue — APPROVED** [review]
👍 Code/PR looks good to me now.  Thanks for all the hard work @KevinVdV and @Raf-atmire .  Just waiting on Travis to succeed before we merge it.

**Pull request merged**
